### PR TITLE
auth refactor

### PR DESCRIPTION
**Pull request by ehuelsmann**
Refactoring authentication in request handling in order to support different authentication models. This commit doesn't change the authentication model itself, but opens up possibilities. Session management and authentication is spread over 3 separate concerns: web sessions (stored in a cookie), authentication and application sessions (stored in the database 'session' table). Each is dealt with in a specific middleware layer.

Note that this PR also changes `old/bin/old-handler.pl` to a module (`old/lib/LedgerSMB/oldHandler.pm`) which allows it to both be pre-loaded as well as function as a PSGI app.

### CODE REVIEW

File: old/lib/LedgerSMB/Form.pm [hunk 1]
```diff
@@ -106,34 +106,22 @@ sub new {
...
+            utf8::decode($self->{$p});
+            utf8::upgrade($self->{$p});
+            $self->{$p} =~ s/\N{NULL}//g;
+        }
+        $self->{nextsub} //= '';
+        $self->{action} //= $self->{nextsub};
     }
-    $self->{nextsub} //= '';
-    $self->{action} //= $self->{nextsub};
     $self->{version}   = "1.8.0-dev";
```
Comment (ylavoie) [inline old/lib/LedgerSMB/Form.pm]: Why the double quotes?

File: lib/LedgerSMB/Middleware/AuthenticateSession.pm [hunk 2]
```diff
@@ -154,26 +136,20 @@ sub call {
...
         }
 
-        $extended_cookie = _verify_session($env->{'lsmb.db'},
-                                           $env->{'lsmb.company'},
-                                           $session);
-        if (! $extended_cookie) {
+        if (! _verify_session($env->{'lsmb.db'},
+                              $env->{'lsmb.company'},
+                              $session)) {
             $dbh->commit;  # potentially log something
```
Comment (ylavoie) [inline lib/LedgerSMB/Middleware/AuthenticateSession.pm]: There will always be something to commit?

File: old/lib/LedgerSMB/Form.pm (hunk 1 above)
Comment (ehuelsmann) [reply old/lib/LedgerSMB/Form.pm]: Good point. No good reason at all. I'll change.

File: lib/LedgerSMB/Middleware/AuthenticateSession.pm (hunk 2 above)
Comment (ehuelsmann) [reply lib/LedgerSMB/Middleware/AuthenticateSession.pm]: I'm not sure. I think at least sometimes: when the session is expired, we want to commit the removal of the session.

Comment (ylavoie) [reply lib/LedgerSMB/Middleware/AuthenticateSession.pm]: Will we generate an error if there’s nothing to commit?

Comment (ehuelsmann) [reply lib/LedgerSMB/Middleware/AuthenticateSession.pm]: Nope. it just commits an empty transaction.

**Comment by ehuelsmann** [comment]
@nick-prater , @ylavoie : the code is in a state that I'm ready to receive comments. Please review; this is quite a step away from our old code structure.

File: old/lib/LedgerSMB/oldHandler.pm [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    # because it changes global state without clean-up
+    #
+    #
+    # Note that this function can receive a PSGI environment, but
+    # modifications aren't marshalled back to the fork()ing process!
+    my ($class, $psgi_env) = @_;
+
+    binmode (STDIN, ':utf8');
+    binmode (STDOUT, ':utf8');
+    binmode STDERR, ':utf8';
```
Comment (nick-prater) [inline old/lib/LedgerSMB/oldHandler.pm]: Inconsistent use of brackets compared with preceding lines.

File: lib/LedgerSMB/Middleware/Authenticate/Company.pm [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+When C<lsmb.want_db> is false (or is missing), a callback C<lsmb.db_cb> is
+added to authenticate and create a database connection on demand.
+
+
+=head1 CALLBACKS
+
+=head2 lsmb.db_cb
+
+###TODOC
```
Comment (nick-prater) [inline lib/LedgerSMB/Middleware/Authenticate/Company.pm]: Documentation needed?

File: old/lib/LedgerSMB/oldHandler.pm (hunk 3 above)
Comment (ehuelsmann) [reply old/lib/LedgerSMB/oldHandler.pm]: You caught me copy/pasting from different sources! (thanks! :-) )

Comment (ehuelsmann) [reply old/lib/LedgerSMB/oldHandler.pm]: Fix submitted.

File: lib/LedgerSMB/Middleware/Authenticate/Company.pm (hunk 4 above)
Comment (ehuelsmann) [reply lib/LedgerSMB/Middleware/Authenticate/Company.pm]: Now documented.

**Pull request merged**